### PR TITLE
Scala 3 DSL

### PR DESCRIPTION
**Pull request by krzemin**
This PR brings complete implementation of the whole Chimney's DSL for Scala 3 (#275) together with moving tests to the shared scala directory.

Due to macro API changes, two DSLs implementations are sharing only minimal set of common types and they are implemented quite independently. However they should be source-compatible.

There is shared test suite (migrated and fixed within this PR) that passes for Scala 2 and compiles for Scala 3.

### CODE REVIEW

File: chimney/src/main/scala-3/io/scalaland/chimney/compiletime/macros/dsl/TransformerDefinitionImpl.scala [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+      From: Type,
+      To: Type,
+      C <: TransformerCfg: Type,
+      F <: TransformerFlags: Type,
+      Inst: Type
+  ](
+      td: Expr[TransformerDefinition[From, To, C, F]],
+      fExpr: Expr[Inst => To]
+  )(using Quotes): Expr[TransformerDefinition[From, To, ? <: TransformerCfg, F]] = {
+    import quotes.reflect.*
```
Comment (jchyb) [inline chimney/src/main/scala-3/io/scalaland/chimney/compiletime/macros/dsl/TransformerDefinitionImpl.scala:82]: NIT: Unnecessary import I believe

File: chimney/src/main/scala-3/io/scalaland/chimney/compiletime/macros/dsl/TransformerDefinitionImpl.scala [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  def buildTransformer[
+      From: Type,
+      To: Type,
+      C <: TransformerCfg: Type,
+      F <: TransformerFlags: Type,
+      ScopeFlags <: TransformerFlags: Type
+  ](
+      td: Expr[TransformerDefinition[From, To, C, F]]
+  )(using Quotes): Expr[Transformer[From, To]] = {
+    import quotes.reflect.*
```
Comment (jchyb) [inline chimney/src/main/scala-3/io/scalaland/chimney/compiletime/macros/dsl/TransformerDefinitionImpl.scala:95]: same here

**Review by jchyb — COMMENTED** [review]
Looks great! I recall also having to move stuff out of the test cases, even after porting to ScalaTest I believe, so that seems unavoidable. Another issue I had with utest happened after expanding the chimney macros, when utest started expanding their own it was crashing on a "Symbol missing owner" in a few testcases, so this one may still appear (although the issue there might have been with my implementation). Anyway, happy I could help with this, even if the one thing that survived from my branch for now is that nightmarish SelectLike structure :) Also feel free to use the TransformerMacros and utils [implementation](https://github.com/jchyb/chimney/blob/scala-3-migration/chimney3/src/main/scala/io/scalaland/internal/macros/TransformerMacros.scala) from there in the future, I tried to make it so that the methods there are aligned with Scala 2 in terms of names and what they do, so whatever refactoring the Scala 2 version implementation will go through, it will be hopefully more or less also applicable there.

**Comment by krzemin** [comment]
> happy I could help with this, even if the one thing that survived from my branch for now is that nightmarish SelectLike structure 

@jchyb thanks a lot! Even if the final code structure doesn't reflect it, believe me - it was much more than just this :)

Regarding field name extraction, there's still a warning that I don't yet know how to fix.

```
[warn] -- Unchecked Warning: chimney/chimney/src/main/scala-3/io/scalaland/chimney/compiletime/utils/DslUtils.scala:61:39 
[warn] 61 |              List(DefDef(_, List(List(ValDef(in, _, _))), _, Some(SelectLike(out, va)))),
[warn]    |                                       ^
[warn]    |the type test for quotes.reflect.ValDef cannot be checked at runtime because it refers to an abstract type member or type parameter
[warn] one warning found
```

**Review by MateuszKubuszok — COMMENTED** [review]
I have a few questions mostly about naming.

File: chimney/src/main/scala-3/io/scalaland/chimney/dsl/TransformerDefinitionCommons.scala [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+private[dsl] trait TransformerDefinitionCommons[UpdateCfg[_ <: TransformerCfg]] {
+
+  import TransformerDefinitionCommons.*
+
+  /** runtime storage for values and functions that transformer definition is customized with */
+  val runtimeData: RuntimeDataStore
+
+  /** updates runtime data in the upper transformer definition  */
+  protected def __updateRuntimeData(newRuntimeData: RuntimeDataStore): this.type
```
[the remark this answers is not in the document]
Comment (krzemin) [reply chimney/src/main/scala-3/io/scalaland/chimney/dsl/TransformerDefinitionCommons.scala:20]: This method is not marked as `inline`, so other inline methods that call this one only inline call to this method in the place when they are called. I assume that control access (protected/private) happens after inline expansion, so no need for extra smartness in order for that to work fine.

File: chimney/src/main/scala-3/io/scalaland/chimney/dsl/PartialTransformerDefinition.scala [hunk 4]
```diff
@@ -2,24 +2,76 @@ package io.scalaland.chimney.dsl
...
 import io.scalaland.chimney.internal.*
+import io.scalaland.chimney.compiletime.macros.dsl.*
 
 /** Allows customization of [[io.scalaland.chimney.PartialTransformer]] derivation.
  *
  * @tparam From  type of input value
  * @tparam To    type of output value
  * @tparam C     type-level encoded config
- * @tparam Flags type-level encoded flags
+ * @tparam F type-level encoded flags
```
[the remark this answers is not in the document]
Comment (krzemin) [reply chimney/src/main/scala-3/io/scalaland/chimney/dsl/PartialTransformerDefinition.scala:12]: Nah,  I was having some weird name clashes with `scala.Flags` when there was `Flags` in Scala 3 🤔 
In most of the places `C` and `F` params are bounded so it should be quite clear what they are.

File: chimney/src/main/scala-3/io/scalaland/chimney/compiletime/macros/dsl/TransformerDefinitionImpl.scala [hunk 5]
```diff
@@ -0,0 +1,97 @@
+package io.scalaland.chimney.compiletime.macros.dsl
```
[the remark this answers is not in the document]
Comment (krzemin) [reply chimney/src/main/scala-3/io/scalaland/chimney/compiletime/macros/dsl/TransformerDefinitionImpl.scala]: why not 👍

**Review by MateuszKubuszok — APPROVED** [review]

**Comment by limpid-kzonix** [comment]
Hey guys. Please provide information about when this will be merged and published as the artifact. Thanks

**Review by MateuszKubuszok — APPROVED** [review]

**Comment by MateuszKubuszok** [comment]
@limpid-kzonix hello. Just to clarify this PR is about _DSL_ part of the migration, currently it still returns stubs rather than derived implementations. The whole progress of the migration can be traced in https://github.com/scalalandio/chimney/milestone/3. We are not committed to any release date yet.

**Pull request merged**